src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+        adapterMethod: () =>
+            adapter.getResourcesByPermissions(
+                resourceType,
+                requiredAccessRoles,
+                searchParams
+            ),
+        refetchDependencies: [adapter, searchParams],
+        isAdapterCalledOnMount: !disabled && shouldFetchResourcesOnMount
+    });
+
+    const placeholder = useMemo(() => {
+        if (displayField === AzureResourceDisplayFields.url) {
+            switch (resourceType) {
+                case AzureResourceTypes.DigitalTwinInstance:
+                    return t('resourcesPicker.enterEnvironmentUrl');
+                case AzureResourceTypes.StorageAccount:
+                    return t('resourcesPicker.enterStorageAccountUrl');
+                case AzureResourceTypes.StorageBlobContainer:
+                    return t('resourcesPicker.enterContainerUrl');
+                default:
+                    return 'resourcesPicker.select';
+            }
+        } else {
+            if (options.length) {
+                return t('resourcesPicker.select');
+            } else {
+                return t('resourcesPicker.noOption');
+            }
+        }
+    }, [resourceType, t, options.length]);
+
+    const loadingLabelText = useMemo(() => {
+        switch (resourceType) {
+            case AzureResourceTypes.DigitalTwinInstance:
+                return t('resourcesPicker.loadingInstances');
+            case AzureResourceTypes.StorageBlobContainer:
+                return t('resourcesPicker.loadingContainers');
+            case AzureResourceTypes.StorageAccount:
+                return t('resourcesPicker.loadingStorageAccounts');
+            default:
+                return t('resourcesPicker.loadingResources');
+        }
+    }, [resourceType, t]);
+
+    const getDisplayFieldValue = useCallback(
+        (resource: IAzureResource) => {
+            if (displayField === AzureResourceDisplayFields.url) {
+                switch (resource.type) {
+                    case AzureResourceTypes.DigitalTwinInstance:
+                        return resource.properties.hostName
+                            ? 'https://' + resource.properties.hostName
+                            : null;
+                    case AzureResourceTypes.StorageAccount:
+                        return resource.properties.primaryEndpoints?.blob;
+                    default:
+                        return resource[
+                            AzureResourceDisplayFields[displayField] // storage blob containers does not have url information in its resource properties, it is constructed based on the storage account blob url
+                        ];
+                }
+            } else {
+                return resource[AzureResourceDisplayFields[displayField]];
+            }
+        },
+        [displayField]
+    );
+
+    // after fetching the resources merge them with existing options in the dropdown
+    const mergeOptionsWithProps = useCallback(
```

Comment:
can this get pulled out to module scope? It seems like it can take everything via props and that would simplify this whole flow of nested functions.

Reply:
it uses resourcesRef inside and one more internal method. will keep it as is